File: apps/android/app/src/main/java/chat/rogi/rogichat/channelport/core/model/song/SongRequest.kt
package chat.rogi.rogichat.channelport.core.model.song

import kotlinx.serialization.Serializable

data class PublicLiveSession(
    val sessionId: Int?,
    val isLive: Boolean,
    val settings: PublicLiveSessionSettings?,
    val queueCount: Int,
)

data class PublicLiveSessionSettings(
    val requestEnabled: Boolean,
    val paused: Boolean,
    val requestCommand: String,
    val maxQueueSize: Int,
    val donationPriorityEnabled: Boolean,
    val chatRequestEnabled: Boolean,
    val donationRequestEnabled: Boolean,
    val requireSongMatch: Boolean,
    val preventDuplicateSongs: Boolean = false,
    val blockedCategoryIds: List<Int> = emptyList(),
    val karaokePlaybackMode: String? = null,
    val karaokeVideoType: String? = null,
    val requestMode: String = "EVERYONE",
)

data class SongRequest(
    val id: Int,
    val liveSessionId: Int,
    val songId: Int? = null,
    val rawArtist: String,
    val rawTitle: String,
    val requesterPlatformId: String,
    val requesterNickname: String,
    val status: String,
    val source: String,
    val queueOrder: Int,
)

@Serializable
data class PublicLiveSessionResponseDto(
    val sessionId: Int? = null,
    val isLive: Boolean = false,
    val settings: PublicLiveSessionSettingsDto? = null,
    val queueCount: Int = 0,
) {
    fun toDomain(): PublicLiveSession = PublicLiveSession(
        sessionId = sessionId,
        isLive = isLive,
        settings = settings?.toDomain(),
        queueCount = queueCount,
    )
}

@Serializable
data class PublicLiveSessionSettingsDto(
    val requestEnabled: Boolean = false,
    val paused: Boolean = false,
    val requestCommand: String = "!신청",
    val maxQueueSize: Int = 0,
    val donationPriorityEnabled: Boolean = true,
    val chatRequestEnabled: Boolean = true,
    val donationRequestEnabled: Boolean = true,
    val requireSongMatch: Boolean = true,
    val preventDuplicateSongs: Boolean = false,
    val blockedCategoryIds: List<Int> = emptyList(),
    val karaokePlaybackMode: String? = null,
    val karaokeVideoType: String? = null,
    val requestMode: String = "EVERYONE",
) {
    fun toDomain(): PublicLiveSessionSettings = PublicLiveSessionSettings(
        requestEnabled = requestEnabled,
        paused = paused,
        requestCommand = requestCommand,
        maxQueueSize = maxQueueSize,
        donationPriorityEnabled = donationPriorityEnabled,
        chatRequestEnabled = chatRequestEnabled,
        donationRequestEnabled = donationRequestEnabled,
        requireSongMatch = requireSongMatch,
        preventDuplicateSongs = preventDuplicateSongs,
        blockedCategoryIds = blockedCategoryIds,
        karaokePlaybackMode = karaokePlaybackMode,
        karaokeVideoType = karaokeVideoType,
        requestMode = requestMode,
    )
}

@Serializable
data class CreateSongRequestPayload(
    val liveSessionId: Int,
    val songId: Int? = null,
    val rawArtist: String,
    val rawTitle: String,
    val rawMessage: String? = null,
    val requesterPlatformId: String,
    val requesterNickname: String,
    val source: String = "MANUAL",
    val donationAmount: Int? = null,
    val donationNativeAmount: Int? = null,
    val donationCurrency: String? = null,
)

@Serializable
data class SongRequestDto(
    val id: Int,
    val liveSessionId: Int,
    val songId: Int? = null,
    val rawArtist: String,
    val rawTitle: String,
    val requesterPlatformId: String,
    val requesterNickname: String,
    val status: String,
    val source: String,
    val queueOrder: Int = 0,
) {
    fun toDomain(): SongRequest = SongRequest(
        id = id,
        liveSessionId = liveSessionId,
        songId = songId,
        rawArtist = rawArtist,
        rawTitle = rawTitle,
        requesterPlatformId = requesterPlatformId,
        requesterNickname = requesterNickname,
        status = status,
        source = source,
        queueOrder = queueOrder,
    )
}

data class SongRequestQueueItem(
    val id: Int,
    val rawArtist: String,
    val rawTitle: String,
    val requesterNickname: String,
    val status: String,
    val queueOrder: Int,
    val songTitle: String?,
    val songArtistName: String?,
    val songAlbumArt: String?,
)

@Serializable
data class SongRequestQueueItemDto(
    val id: Int,
    val rawArtist: String,
    val rawTitle: String,
    val requesterNickname: String,
    val status: String,
    val queueOrder: Int = 0,
    val song: SongRequestSongDto? = null,
) {
    fun toDomain(): SongRequestQueueItem = SongRequestQueueItem(
        id = id,
        rawArtist = rawArtist,
        rawTitle = rawTitle,
        requesterNickname = requesterNickname,
        status = status,
        queueOrder = queueOrder,
        songTitle = song?.title,
        songArtistName = song?.artist?.name,
        songAlbumArt = song?.albumArt,
    )
}

@Serializable
data class SongRequestSongDto(
    val id: Int,
    val title: String,
    val artist: SongRequestArtistDto? = null,
    val albumArt: String? = null,
)

@Serializable
data class SongRequestArtistDto(
    val id: Int,
    val name: String,
)

@Serializable
data class RequestedSongIdsResponseDto(
    val songIds: List<Int> = emptyList(),
)

@Serializable
data class SongRequestQueueResponseDto(
    val requests: List<SongRequestQueueItemDto> = emptyList(),
    val total: Int = 0,
)

File: apps/ios/Sources/Features/Channel/SongRequest.swift
import Foundation

// MARK: - Domain Models

struct LiveSession {
    let sessionId: Int?
    let isLive: Bool
    let settings: LiveSessionSettings?
    let queueCount: Int
}

enum SongRequestMode: String, Decodable {
    case everyone = "EVERYONE"
    case verifiedOnly = "VERIFIED_ONLY"
    case chatOnly = "CHAT_ONLY"
}

struct LiveSessionSettings {
    let requestEnabled: Bool
    let chatRequestEnabled: Bool
    let donationRequestEnabled: Bool
    let paused: Bool
    let maxQueueSize: Int
    let preventDuplicateSongs: Bool
    let blockedCategoryIds: [Int]
    let requestMode: SongRequestMode
}

struct SongRequestPayload: Encodable {
    let liveSessionId: Int
    let songId: Int?
    let rawArtist: String
    let rawTitle: String
    let rawMessage: String?
    let requesterPlatformId: String
    let requesterNickname: String
    let source: String = "MANUAL"
}

// MARK: - DTOs

struct LiveSessionDTO: Decodable {
    let sessionId: Int?
    let isLive: Bool?
    let settings: LiveSessionSettingsDTO?
    let queueCount: Int?

    func toDomain() -> LiveSession {
        LiveSession(
            sessionId: sessionId,
            isLive: isLive ?? false,
            settings: settings?.toDomain(),
            queueCount: queueCount ?? 0
        )
    }
}

struct LiveSessionSettingsDTO: Decodable {
    let requestEnabled: Bool?
    let chatRequestEnabled: Bool?
    let donationRequestEnabled: Bool?
    let paused: Bool?
    let maxQueueSize: Int?
    let preventDuplicateSongs: Bool?
    let blockedCategoryIds: [Int]?
    let requestMode: String?

    func toDomain() -> LiveSessionSettings {
        LiveSessionSettings(
            requestEnabled: requestEnabled ?? false,
            chatRequestEnabled: chatRequestEnabled ?? true,
            donationRequestEnabled: donationRequestEnabled ?? true,
            paused: paused ?? false,
            maxQueueSize: maxQueueSize ?? 0,
            preventDuplicateSongs: preventDuplicateSongs ?? false,
            blockedCategoryIds: blockedCategoryIds ?? [],
            requestMode: SongRequestMode(rawValue: requestMode ?? "EVERYONE") ?? .everyone
        )
    }
}

struct RequestedSongIdsResponse: Decodable {
    let songIds: [Int]
}

struct SongRequestResponse: Decodable {
    let id: Int
    let liveSessionId: Int
    let songId: Int?
    let rawArtist: String
    let rawTitle: String
    let status: String
    let queueOrder: Int?
}

// MARK: - Song Request Queue

struct SongRequestQueueItem: Identifiable {
    let id: Int
    let rawArtist: String
    let rawTitle: String
    let requesterNickname: String
    let status: String
    let queueOrder: Int
    let songTitle: String?
    let songArtistName: String?
    let songAlbumArt: String?

    var displayTitle: String { songTitle ?? rawTitle }
    var displayArtist: String { songArtistName ?? rawArtist }

    var statusLabel: String {
        switch status {
        case "PENDING": return "대기"
        case "ACCEPTED": return "수락"
        case "PLAYING": return "재생"
        default: return status
        }
    }
}

struct SongRequestQueueItemDTO: Decodable {
    let id: Int
    let rawArtist: String
    let rawTitle: String
    let requesterNickname: String
    let status: String
    let queueOrder: Int?
    let song: SongRequestSongDTO?

    func toDomain() -> SongRequestQueueItem {
        SongRequestQueueItem(
            id: id,
            rawArtist: rawArtist,
            rawTitle: rawTitle,
            requesterNickname: requesterNickname,
            status: status,
            queueOrder: queueOrder ?? 0,
            songTitle: song?.title,
            songArtistName: song?.artist?.name,
            songAlbumArt: song?.albumArt
        )
    }
}

struct SongRequestSongDTO: Decodable {
    let id: Int
    let title: String
    let artist: SongRequestArtistDTO?
    let albumArt: String?
}

struct SongRequestArtistDTO: Decodable {
    let id: Int
    let name: String
}

struct SongRequestQueueResponseDTO: Decodable {
    let requests: [SongRequestQueueItemDTO]
    let total: Int
}
